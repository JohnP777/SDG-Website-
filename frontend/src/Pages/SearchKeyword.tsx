import Page from '../Components/Page';
import { Box, Typography } from '@mui/material';
import { useEffect, useState } from 'react';
import teamsImg from '../Assets/teams.jpg';
import actionformImg from '../Assets/actionform.jpg';
import databaseImg from '../Assets/database.jpg';

const SearchKeyword = () => {
  const [loggedIn, setLoggedIn] = useState(false);
  useEffect(() => {
    const token = localStorage.getItem('token');
    const expiry = localStorage.getItem('token-expiry');
    if (token && expiry && Date.now() < new Date(expiry).getTime()) {
      setLoggedIn(true);
    } else {
      setLoggedIn(false);
    }
  }, []);

  const teamsLink = loggedIn ? '/teams' : '/signup';
  const actionFormLink = loggedIn ? '/sdg-form' : '/signup';

  return (
    <Page>
      <Box sx={{
        display: 'flex',
        flexDirection: 'row',
        justifyContent: 'center',
        alignItems: 'center',
        gap: { xs: 2, md: 8 },
        marginTop: '10px',
        minHeight: 'calc(100vh - 300px)',
        width: '100%'
      }}>
        {/* Teams */}
        <Box sx={{ display: 'flex', flexDirection: 'column', alignItems: 'center', flex: 1 }}>
          <a href={teamsLink} style={{ textDecoration: 'none' }}>
            <Typography variant="h5" sx={{ color: '#1565c0', fontWeight: 'bold', textAlign: 'center', mb: 2 }}>
              SDG Action Teams
            </Typography>
            <img src={teamsImg} alt="Teams" style={{ width: '100%', maxWidth: 340, height: 260, objectFit: 'cover', borderRadius: 12, boxShadow: '0 2px 8px rgba(0,0,0,0.08)', cursor: 'pointer' }} />
          </a>
        </Box>
        {/* Action Form */}
        <Box sx={{ display: 'flex', flexDirection: 'column', alignItems: 'center', flex: 1 }}>
          <a href={actionFormLink} style={{ textDecoration: 'none' }}>
            <Typography variant="h5" sx={{ color: '#1565c0', fontWeight: 'bold', textAlign: 'center', mb: 2 }}>
              Create an SDG Action Plan
            </Typography>
            <img src={actionformImg} alt="Action Form" style={{ width: '100%', maxWidth: 340, height: 260, objectFit: 'cover', borderRadius: 12, boxShadow: '0 2px 8px rgba(0,0,0,0.08)', cursor: 'pointer' }} />
          </a>
        </Box>
        {/* Database */}
        <Box sx={{ display: 'flex', flexDirection: 'column', alignItems: 'center', flex: 1 }}>
          <a href="https://sdg.unswzoo.com/" target="_blank" rel="noopener noreferrer" style={{ textDecoration: 'none' }}>
            <Typography variant="h5" sx={{ color: '#1565c0', fontWeight: 'bold', textAlign: 'center', mb: 2 }}>
              SDG Database
            </Typography>
            <img src={databaseImg} alt="Database" style={{ width: '100%', maxWidth: 340, height: 260, objectFit: 'cover', borderRadius: 12, boxShadow: '0 2px 8px rgba(0,0,0,0.08)', cursor: 'pointer' }} />
          </a>
        </Box>
      </Box>
    </Page>
  );
};

export default SearchKeyword;
