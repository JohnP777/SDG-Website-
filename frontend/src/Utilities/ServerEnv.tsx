export const getUrl = () => {
    if (process.env.REACT_APP_BACKEND_URL) {
        return process.env.REACT_APP_BACKEND_URL;
    }
    return 'http://127.0.0.1:8000/'
}