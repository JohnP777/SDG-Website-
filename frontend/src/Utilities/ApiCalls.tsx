import { getUrl } from "./ServerEnv";

export const apiCallPost = async (path: string, body: any, isAuthed: boolean ) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'POST',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
      body: JSON.stringify(body)
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallGet = async (path: string, isAuthed: boolean, headersExtra?: any) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'GET',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : '',
        ...headersExtra
      },
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallPatch = async (path: string, body: object, isAuthed: boolean) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'PATCH',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
      body: JSON.stringify(body)
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallDelete = async (path: string, isAuthed: boolean) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'DELETE',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallPut = async (path: string, body: any, isAuthed: boolean) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'PUT',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
      body: JSON.stringify(body)
    });
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error);
  }
};
