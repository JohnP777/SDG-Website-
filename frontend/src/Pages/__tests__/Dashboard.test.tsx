import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import SearchKeyword from "../SearchKeyword";
import * as ApiCalls from "../../Utilities/ApiCalls"
import { MemoryRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { GoogleOAuthProvider } from '@react-oauth/google';

// Mock the API calls
jest.mock("../../Utilities/ApiCalls", () => ({
  apiCallGet: jest.fn(),
  apiCallPost: jest.fn()
}));

// Mock window.location.reload
const mockReload = jest.fn();
Object.defineProperty(window, 'location', {
  value: {
    reload: mockReload,
  },
  writable: true
});

// Mock useNavigate
const mockedUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));

// Helper function to render the component with router
const renderWithRouter = (component: React.ReactNode) => {
  return render(
    <GoogleOAuthProvider clientId="google-client-id">
      <MemoryRouter>
        {component}
      </MemoryRouter>
    </GoogleOAuthProvider>
  );
};

describe("CreateDashboard/Search", () => {
  beforeEach(() => {
    jest.clearAllMocks();

    (ApiCalls.apiCallPost as jest.Mock).mockImplementation((url) => {
      if (url === 'api/admin/log/education/' || url === 'api/admin/log/action/') {
        return Promise.resolve({ statusCode: 200 });
      } else if (url === 'api/admin/analytics/educations/top/') {
        return Promise.resolve({
          0: {
            educationName: 'Introducing the Sustainability Bootcamp and the SDGs',
          }
        });
      } else if (url === 'api/admin/analytics/actions/top/') {
        return Promise.resolve({
          0: {
            actionName: "Get rid of paper bank statements",
          }
        });
      }
    });

    (ApiCalls.apiCallGet as jest.Mock).mockImplementation((url) => {
      if (url.includes('userInteractions')) {
        return Promise.resolve({
          action_plans_viewed: [],
          education_plans_viewed: [],
        });
      } else if (url.startsWith('api/sdg-education/search?q=Introducing')) {
        return Promise.resolve([
          {
            title: "Introducing the Sustainability Bootcamp and the SDGs",
            aims: "Understand transboundary freshwater governance.",
            descriptions: "Course on governance for transboundary freshwater security...",
            organization: "SDG Academy X"
          }
        ]);
      } else if (url.startsWith('api/sdg-actions/search?q=Get')) {
        return Promise.resolve([
          {
            actions: "Get rid of paper bank statements",
            action_detail: "Cut down on waste and maybe save money at the coffee shop.",
          }
        ]);
      }
      return Promise.resolve([]);
    });

    // user isn't logged in
    Storage.prototype.getItem = jest.fn((key) => {
      if (key === 'userDetails') {
        return JSON.stringify({ id: '1', name: 'Test User' });
      }
      if (key === 'token') {
        return 'fake_token';
      }
      return null;
    });
  });

  it("should match snapshot", async () => {
    const { asFragment } = renderWithRouter(<SearchKeyword />);
    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it("should render the title", async () => {
    renderWithRouter(<SearchKeyword />);
    expect(await screen.findByText("SDG Knowledge System")).toBeInTheDocument();
  });

  it("should display recent searches when the input is focused", async () => {
    renderWithRouter(<SearchKeyword />);
    const input = screen.getByPlaceholderText("Search for an SDG Action/Education Plan");
    fireEvent.focus(input);

    expect(await screen.findByText("Get rid of paper bank statements")).toBeInTheDocument();
    expect(await screen.findByText("Introducing the Sustainability Bootcamp and the SDGs")).toBeInTheDocument();
  });

  it("should fetch suggestions when typing", async () => {
    renderWithRouter(<SearchKeyword />);
    const input = screen.getByPlaceholderText("Search for an SDG Action/Education Plan");
    
    fireEvent.focus(input);
    fireEvent.change(input, { target: { value: "Introducing" } });

    await waitFor(() => {
      expect(ApiCalls.apiCallGet).toHaveBeenCalledWith(
        expect.stringContaining("api/sdg-education/search?q=Introducing"),
        false
      );
      expect(ApiCalls.apiCallGet).toHaveBeenCalledWith(
        expect.stringContaining("api/sdg-actions/search?q=Introducing"),
        false
      );
    });

  });

  it("should update search input value", async () => {
    renderWithRouter(<SearchKeyword />);
    const input = screen.getByPlaceholderText("Search for an SDG Action/Education Plan");

    userEvent.type(input, "Sustainability");

    await waitFor(() => {
      expect(input).toHaveValue("Sustainability");
    });
  });

  it("should open modal when suggested plan is selected", async () => {
    renderWithRouter(<SearchKeyword />);
    const input = screen.getByPlaceholderText("Search for an SDG Action/Education Plan");
    
    fireEvent.focus(input);
    await userEvent.type(input, "Introducing");
  
    const suggestion = await screen.findByText("Introducing the Sustainability Bootcamp and the SDGs");
    expect(suggestion).toBeInTheDocument();
  
    userEvent.click(suggestion);
  });

});
