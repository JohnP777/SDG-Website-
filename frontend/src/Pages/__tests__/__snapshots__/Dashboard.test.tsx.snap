// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateDashboard/Search should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-z3339j"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-u2vtxm-MuiPaper-root-MuiAppBar-root"
      style="--Paper-shadow: none;"
    >
      <div
        class="MuiContainer-root css-12pwiy3-MuiContainer-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular css-1t0x8et-MuiToolbar-root"
        >
          <a
            class="MuiTypography-root MuiTypography-subtitle1 css-1z0p208-MuiTypography-root"
            href="/"
          >
            SDG
            <br />
            ZOO
          </a>
          <div
            class="MuiBox-root css-dq25au"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge css-z77o6z-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1xf0eoi-MuiSvgIcon-root"
                data-testid="MenuIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3zm0-5h18v-2H3zm0-7v2h18V6z"
                />
              </svg>
            </button>
          </div>
          <div
            class="MuiBox-root css-10s323d"
          >
            <a
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1oyog3l-MuiButtonBase-root-MuiButton-root"
              href="/"
              tabindex="0"
            >
              HOME
            </a>
            <a
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1oyog3l-MuiButtonBase-root-MuiButton-root"
              href="/about-us"
              tabindex="0"
            >
              ABOUT
            </a>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-769nkz-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              FEATURES
              <span
                class="MuiButton-icon MuiButton-endIcon MuiButton-iconSizeMedium css-1gulhci-MuiButton-endIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="KeyboardArrowDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            class="MuiBox-root css-1t0e2o"
          >
            <button
              aria-label="Profile settings"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1avsmgh-MuiButtonBase-root-MuiIconButton-root"
              data-mui-internal-clone-element="true"
              tabindex="0"
              type="button"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-bqh3fr-MuiAvatar-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="PersonIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4m0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4"
                  />
                </svg>
              </div>
            </button>
          </div>
        </div>
      </div>
    </header>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-eyrs0v-MuiContainer-root"
    >
      <div
        class="MuiBox-root css-1ahbek6"
      >
        <h3
          class="MuiTypography-root MuiTypography-h3 css-1sjcxh8-MuiTypography-root"
        >
          SDG Knowledge System
        </h3>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-2v5rmj-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart css-ot336b-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-p9vzyp-MuiSvgIcon-root"
              data-testid="SearchIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
              />
            </svg>
            <input
              aria-invalid="false"
              autocomplete="off"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart css-2u11ia-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r6:"
              placeholder="Search for an SDG Action/Education Plan"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-w4cd9x"
              >
                <span
                  aria-hidden="true"
                  class="notranslate"
                >
                  â€‹
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
